cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -339,22 +385,49 @@ public void apply() throws Exception {
    * @throws Exception
    */
   public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
-        }
-      });
+    throws InterruptedException, IOException, DatasetManagementException {
+    LOG.info("Starting upgrade of DatasetBasedStreamSizeScheduleStore.");
+
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.MINUTES.sleep(1);
+    }
+
+    while (!appVerUpgrade.get()) {
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
+          .execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() {
+              if (upgradeVersionKeys(table)) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'), APP_VERSION_UPGRADE_KEY);
```

Comment:
don't you also want to read this at the very beginning? If it is already set, then there is no need to try and upgrade more.